File: Python/NetworkFlow.py
from collections import deque

INF=10**9+7

def EdmondsKarp(st,ed):
    ret=0
    while True:
        visited=[-1]*(N+1)
        q=deque()
        q.append(st)
        visited[st]=st
        while q:
            x=q.popleft()
            for dx in arr[x]:
                if visited[dx]==-1 and capacity[x][dx]>flow[x][dx]:
                    q.append(dx)
                    visited[dx]=x
        if visited[ed]==-1:
            break
        mn=INF
        i=ed
        while i!=st:
            j=visited[i]
            mn=min(mn,capacity[j][i]-flow[j][i])
            i=j
        i=ed
        while i!=st:
            j=visited[i]
            flow[j][i]+=mn
            flow[i][j]-=mn
            i=j
        ret+=mn
    return ret

N,M,S,E=map(int, input().split())
capacity=[[0]*(N+1) for _ in range(N+1)]
flow=[[0]*(N+1) for _ in range(N+1)]
arr=[[] for _ in range(N+1)]
for _ in range(M):
    a,b,c=map(int, input().split())
    capacity[a][b]+=c
    capacity[b][a]+=c
    arr[a].append(b)
    arr[b].append(a)
print(EdmondsKarp(S,E))

File: Java/NetworkFlow.java
import java.util.*;
@SuppressWarnings("unchecked")

public class Main {
    static final int INF=1000000007;
    static int N, M, S, E;
    static int[][] capacity=new int[1005][1005];
    static int[][] flow=new int[1005][1005];
    static int[] visited=new int[1005];
    static List<Integer>[] arr=new ArrayList[1005];

    static int EdmondsKarp(int st, int ed){
        int ret=0;
        while (true){
            Arrays.fill(visited,-1);
            Queue<Integer> q=new LinkedList<>();
            q.add(st);
            visited[st]=st;
            while (!q.isEmpty()){
                int x=q.poll();
                for (int dx:arr[x]){
                    if (visited[dx]==-1 && capacity[x][dx]>flow[x][dx]){
                        q.add(dx);
                        visited[dx]=x;
                    }
                }
            }
            if (visited[ed]==-1)
                break;
            int mn=INF;
            for (int i=ed, j;i!=st;i=j){
                j=visited[i];
                mn=Math.min(mn,capacity[j][i]-flow[j][i]);
            }
            for (int i=ed, j;i!=st;i=j){
                j=visited[i];
                flow[j][i]+=mn;
                flow[i][j]-=mn;
            }
            ret+=mn;
        }
        return ret;
    }

    public static void main(String[] args){
        Scanner sc=new Scanner(System.in);
        N=sc.nextInt();
        M=sc.nextInt();
        S=sc.nextInt();
        E=sc.nextInt();
        for (int i=1;i<=N;i++)
            arr[i]=new ArrayList<>();
        for (int i=0;i<M;i++){
            int a=sc.nextInt();
            int b=sc.nextInt();
            int c=sc.nextInt();
            capacity[a][b]+=c;
            capacity[b][a]+=c;
            arr[a].add(b);
            arr[b].add(a);
        }
        System.out.println(EdmondsKarp(S, E));
        sc.close();
    }
}
